### ISSUE TITLE
Derivative of pow(x, y) can give nan when x=0 for higher order derivatives

### ISSUE THREAD

**Issue by mjwatkins2** [tensorflow/tensorflow#35011]
**System information**
- Have I written custom code: Example provided below
- OS Platform and Distribution: Both Windows 10 and Google Colab
- TensorFlow installed from binary
- TensorFlow version: Both 2.0.0 and 2.1.0-rc0
- Python version: 3.7.4

**Describe the current behavior**
The output of pow(x, y) is not correct for higher order derivatives when x is 0 and when y is other than a single value. For case 2 below I get:
[0 0 4]
[[-2 1 10]]
[[8 -nan 20]]
[[-18 -nan 30]]
[[24 -nan 24]]
[[0 -nan 0]]

For case 1 the nan only shows up in the 5th derivative, which is still not great, but not as bad:
[1 0 1]
[[-4 0 4]]
[[12 0 12]]
[[-24 0 24]]
[[24 24 24]]
[[-0 -nan 0]]

**Describe the expected behavior**
The output of this example for case 1 should be:
[1 0 1]
[[-4 0 4]]
[[12 0 12]]
[[-24 0 24]]
[[24 24 24]]
[[-0 0 0]]

The output of this example for case 2 should be:
[0 0 4]
[[-2 1 10]]
[[8 0 20]]
[[-18 0 30]]
[[24 0 24]]
[[0 0 0]]

**Code to reproduce the issue**

```
import tensorflow as tf

x = tf.Variable([[-1.0], [0.0], [1.0]])

with tf.GradientTape(persistent=True) as t:
    t.watch(x)
    # case 1: y = x^4
    # y = tf.reduce_sum(tf.pow(x, 4), axis=1) # gives nan for 5th derivative at x=0
    # case 2: y = x + x^2 + x^3 + x^4
    y = tf.reduce_sum(tf.pow(x, [[1, 2, 3, 4]]), axis=1) # gives nan for 2nd to 5th derivative at x=0
    dy_dx = t.gradient(y, x)
    d2y_dx2 = t.gradient(dy_dx, x)
    d3y_dx3 = t.gradient(d2y_dx2, x)
    d4y_dx4 = t.gradient(d3y_dx3, x)
    d5y_dx5 = t.gradient(d4y_dx4, x)
del t

tf.print(y)
tf.print(tf.transpose(dy_dx)) # transpose only to fit on one line when printed
tf.print(tf.transpose(d2y_dx2))
tf.print(tf.transpose(d3y_dx3))
tf.print(tf.transpose(d4y_dx4))
tf.print(tf.transpose(d5y_dx5))
```

**Comment by oanush** [comment]
Issue replicating in for given code, kindly find the [gist](https://colab.sandbox.google.com/gist/oanush/0fb369fd96e8df3fbc789b8e662e1176/35011.ipynb) of colab.Thanks!

**Comment by allenlavoie** [comment]
Thank you for the report. I'm checking in a change that defines a gradient for 0 ** 0 wrt the base.

The higher-order gradient examples will still have NaNs. That's an interesting discussion we've had in the past about whether to "fix" backprop in cases like this by not propagating NaNs when multiplying by a zero derivative (so x ** 0 could "hide" NaNs from x ** -1). The conclusion so far has been that we should leave backprop dumb.

**Comment by mjwatkins2** [comment]
I realize that it's preferable to avoid handling too many special cases, but it seems like some value/utility is lost here by not producing the correct higher-order gradients. The actual higher-order gradients are defined and are continuous, but those computed by TF have a discontinuity (at x = 0). It may be unlikely that an exact value of 0 will show up during typical training, but I ran into it and it sounds like it's happened before. What's the rationale on leaving backprop "dumb"?
Thanks!

**Comment by allenlavoie** [comment]
@rmlarsen do you have more background on why we decided not to hide NaNs in backprop?

**Comment by allenlavoie** [comment]
Reopening since I'm rolling back the fix (needs investigation).

**Comment by sushreebarsa** [comment]
Was able to replicate the issue in TF v2.5 ,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/1128ca9668d3fc0452e703fe7625367e/untitled118.ipynb)..Thanks !

**Comment by kumariko** [comment]
I could reproduce the issue with TF 2.6 . Please, find the [**`gist`**](https://colab.research.google.com/gist/kumariko/d2332d8c553b5afe2e714b2f3431d244/untitled118.ipynb#scrollTo=LR8L4OwgXAfr) here.Thanks!

**Comment by chunduriv** [comment]
I was able to replicate the issue in tf-nightly 2.10.0-dev20220726. Please find the [gist](https://colab.sandbox.google.com/gist/chunduriv/c1c8b7f464c8a1c7e6d5d4b256684aac/untitled45.ipynb) for reference. Thank you.

**Comment by Venkat6871** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
